Reject empty groups

bootstraponline opened the pull request:
This seems like a simple bug. If I have two files (one spec in each file) then I expect 1 spec per process (2 processes for 2 specs). The current behavior is 4 processes which doesn't make sense.

Before:
```
$ parallel_rspec spec/
4 processes for 2 specs, ~ 0 specs per process
```

After:
```
$ parallel_rspec spec/
2 processes for 2 specs, ~ 1 specs per process
```

bootstraponline commented:
Tests are failing on JRuby, I don't think it's related to my pull request.

grosser commented on lib/parallel_tests/cli.rb:
```diff
@@ -35,6 +35,7 @@ def run_tests_in_parallel(num_processes, options)
+        groups.reject! &:empty?
 
         test_results = if options[:only_group]
           groups_to_run = options[:only_group].collect{|i| groups[i - 1]}
```
that would blow up if the index no longer exists

bootstraponline replied:
It returns nil on invalid index. I'll update to use compact.

```ruby
[16] pry(#<ParallelTests::CLI>)> [1,2,8].collect { |i| groups[i-1] }
=> [["spec/two_spec.rb"], ["spec/one_spec.rb"], nil]
[17] pry(#<ParallelTests::CLI>)> [1,2,8].collect { |i| groups[i-1] }.compact
=> [["spec/two_spec.rb"], ["spec/one_spec.rb"]]
[18] pry(#<ParallelTests::CLI>)> [1,2,8].collect { |i| [][i-1] }.compact
=> []
[19] pry(#<ParallelTests::CLI>)> [1,2,8].collect { |i| [][i-1] }
=> [nil, nil, nil]
```

grosser commented:
I like the idea. Can you add a test that covers it + one that tries to use an index of an empty group -> runs  nothing.

bootstraponline commented:
Sure, I will work on updating this.

bootstraponline commented:
> Can you add a test that covers it + one that tries to use an index of an empty group -> runs nothing.

I added the two tests. Let me know if any other changes are required.

bootstraponline commented:
I added Ruby 2.2.2 to Travis since that's the latest stable release. Rails is failing there, appears unrelated to this pull request. Tests are passing on Ruby 1.9, 2.0, 2.1.

grosser commented on lib/parallel_tests/cli.rb:
```diff
@@ -92,8 +93,9 @@ def report_results(test_results)
       num_processes = groups.size
-      num_tests = groups.map(&:size).inject(:+)
-      puts "#{num_processes} processes for #{num_tests} #{name}s, ~ #{num_tests / groups.size} #{name}s per process"
+      num_tests = groups.empty? ? 0 : groups.map(&:size).inject(:+)
```
`.inject(0, :+)` is a bit nicer :)

bootstraponline replied:
I agree, fixed.

grosser commented on lib/parallel_tests/cli.rb:
```diff
@@ -92,8 +93,9 @@ def report_results(test_results)
-      num_tests = groups.map(&:size).inject(:+)
-      puts "#{num_processes} processes for #{num_tests} #{name}s, ~ #{num_tests / groups.size} #{name}s per process"
+      num_tests = groups.empty? ? 0 : groups.map(&:size).inject(:+)
+      tests_per_process = num_processes <= 0 ? 0 : num_tests / num_processes
```
-1 processes seems a bit scary / I'd want that to show :D 
how about:
`tests_per_process = (num_processes == 0 ? 0 : num_tests / num_processes)`

bootstraponline replied:
Fixed

grosser commented on lib/parallel_tests/cli.rb:
```diff
@@ -35,9 +35,10 @@ def run_tests_in_parallel(num_processes, options)
 
         test_results = if options[:only_group]
-          groups_to_run = options[:only_group].collect{|i| groups[i - 1]}
+          groups_to_run = options[:only_group].collect{|i| groups[i - 1]}.compact
```
hmmm nah that is still not 'perfect' ... if there is [[],[t1],[t2]] then running --only-group 1 would run group 2 :(

grosser replied:
I guess that would never happen since we fill index 0 first

grosser replied:
either way good enough for that silly edge case :)

bootstraponline replied:
If there's a better way I'm happy to fix. I think the current code works (at least the tests pass).

grosser commented on spec/parallel_tests/cli_spec.rb:
```diff
@@ -180,6 +180,12 @@ def call(*args)
       end
 
+      it "run no group when only_group uses invalid index" do
+        group_number = 99
```
that's kind of a counter usecase since the group number should never be greater then process number ...
os --only-group 3 good enough for this test ?

-> `runs nothing when --only-group references an empty group`

bootstraponline replied:
I updated it to use a group number of 3.

grosser commented:
rails 3.2 + ruby 2.2.2 don't play nice ... just don't test it on 2.2.2 I guess ...

bootstraponline commented:
I added a new commit which addresses all the comments. Let me know if there are additional changes required. If it looks good then I'll squash to one commit before merging. The tests are passing now on all the Rubies (except JRuby).

grosser commented:
thx! Looks great :)

grosser commented:
1.6.1 is out! thx! :D

bootstraponline commented:
thanks!
